pennylane/optimize/shot_adaptive.py

```diff
@@ -0,0 +1,460 @@
+        device = qnode.device
+
+        self.check_device(qnode.device)
+        original_shots = device.shots
+
+        if self.lipschitz is None:
+            self.check_learning_rate(1)
+
+        try:
+            device.shots = [(1, self.max_shots)]
+            grads = [qml.jacobian(qnode, argnum=i)(*args, **kwargs) for i in self.trainable_args]
+        finally:
+            device.shots = original_shots
+
+        return grads
+
+    def compute_grad(
+        self, objective_fn, args, kwargs
+    ):  # pylint: disable=signature-differs,arguments-differ
+        r"""Compute gradient of the objective function, as well as the variance of the gradient,
+        at the given point.
+
+        Args:
+            objective_fn (function): the objective function for optimization
+            args: arguments to the objective function
+            kwargs: keyword arguments to the objective function
+
+        Returns:
+            tuple[array[float], array[float]]: a tuple of NumPy arrays containing the gradient
+            :math:`\nabla f(x^{(t)})` and the variance of the gradient.
+        """
+        if isinstance(objective_fn, qml.ExpvalCost):
+            grads = self._single_shot_expval_gradients(objective_fn, args, kwargs)
+        elif isinstance(objective_fn, qml.tape.QNode) or hasattr(objective_fn, "device"):
+            grads = self._single_shot_qnode_gradients(objective_fn, args, kwargs)
+        else:
+            raise ValueError(
+                "The objective function must either be encoded as a single QNode or "
+                "an ExpvalCost object for the Rosalin optimizer. "
+            )
+
+        # grads will have dimension [max(self.s), *params.shape]
+        # For each parameter, we want to truncate the number of shots to self.s[idx],
+        # and take the mean and the variance.
+        gradients = []
+        gradient_variances = []
+
+        for i, grad in enumerate(grads):
+            p_ind = np.ndindex(*grad.shape[1:])
+
+            g = np.zeros_like(grad[0])
+            s = np.zeros_like(grad[0])
+
+            for idx in p_ind:
+                grad_slice = grad[(slice(0, self.s[i][idx]),) + idx]
+                g[idx] = np.mean(grad_slice)
+                s[idx] = np.var(grad_slice, ddof=1)
+
+            gradients.append(g)
+            gradient_variances.append(s)
+
+        return gradients, gradient_variances
+
+    def step(self, objective_fn, *args, **kwargs):
+        """Update trainable arguments with one step of the optimizer.
+
+        Args:
+            objective_fn (function): the objective function for optimization
+            *args : Variable length argument list for objective function
+            **kwargs : variable length of keyword arguments for the objective function
+
+        Returns:
+            list[array]: the new variable values :math:`x^{(t+1)}`.
+            If single arg is provided, list [array] is replaced by array.

```

Comment:
Only changed some capitalization and removed a few too many spaces (and added a tab).
```suggestion
            *args: variable length argument list for objective function
            **kwargs: variable length of keyword arguments for the objective function

        Returns:
            list[array]: The new variable values :math:`x^{(t+1)}`.
                If single arg is provided, list[array] is replaced by array.

```

Reply:
Thanks Theo, I edited your suggestion before accepting though, since the `Returns:` section doesn't use indents for continuation lines (this results in Sphinx rendering issues).

This is because the `Returns:` section only has at most one item (you can only return once from a Python function). In contrast, `Args:` will have multiple items, so indentation is required.